In [2]:
import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

In [3]:
os.chdir("/Users/darylldalmeida/Documents/Data science/Datasets/Bank loan")

In [4]:
df = pd.read_csv("Creditcardfraud.csv")

In [5]:
df.head(10)

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815      170136.00       160296.36   
1     1   PAYMENT   1864.28  C1666544295       21249.00        19384.72   
2     1  TRANSFER    181.00  C1305486145         181.00            0.00   
3     1  CASH_OUT    181.00   C840083671         181.00            0.00   
4     1   PAYMENT  11668.14  C2048537720       41554.00        29885.86   
5     1   PAYMENT   7817.71    C90045638       53860.00        46042.29   
6     1   PAYMENT   7107.77   C154988899      183195.00       176087.23   
7     1   PAYMENT   7861.64  C1912850431      176087.23       168225.59   
8     1   PAYMENT   4024.36  C1265012928        2671.00            0.00   
9     1     DEBIT   5337.77   C712410124       41720.00        36382.23   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0            0.00        0               0  
1  M2044282225             0.0            0.00        0               0  
2   C553264065             0.0            0.00        1               0  
3    C38997010         21182.0            0.00        1               0  
4  M1230701703             0.0            0.00        0               0  
5   M573487274             0.0            0.00        0               0  
6   M408069119             0.0            0.00        0               0  
7   M633326333             0.0            0.00        0               0  
8  M1176932104             0.0            0.00        0               0  
9   C195600860         41898.0        40348.79        0               0

In [6]:
df.shape

(6362620, 11)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


## EDA

In [8]:
df.describe()

step        amount  oldbalanceOrg  newbalanceOrig  \
count  6.362620e+06  6.362620e+06   6.362620e+06    6.362620e+06   
mean   2.433972e+02  1.798619e+05   8.338831e+05    8.551137e+05   
std    1.423320e+02  6.038582e+05   2.888243e+06    2.924049e+06   
min    1.000000e+00  0.000000e+00   0.000000e+00    0.000000e+00   
25%    1.560000e+02  1.338957e+04   0.000000e+00    0.000000e+00   
50%    2.390000e+02  7.487194e+04   1.420800e+04    0.000000e+00   
75%    3.350000e+02  2.087215e+05   1.073152e+05    1.442584e+05   
max    7.430000e+02  9.244552e+07   5.958504e+07    4.958504e+07   

       oldbalanceDest  newbalanceDest       isFraud  isFlaggedFraud  
count    6.362620e+06    6.362620e+06  6.362620e+06    6.362620e+06  
mean     1.100702e+06    1.224996e+06  1.290820e-03    2.514687e-06  
std      3.399180e+06    3.674129e+06  3.590480e-02    1.585775e-03  
min      0.000000e+00    0.000000e+00  0.000000e+00    0.000000e+00  
25%      0.000000e+00    0.000000e+00  0.000000e+00    0.000000e+00  
50%      1.327057e+05    2.146614e+05  0.000000e+00    0.000000e+00  
75%      9.430367e+05    1.111909e+06  0.000000e+00    0.000000e+00  
max      3.560159e+08    3.561793e+08  1.000000e+00    1.000000e+00

In [9]:
df['type'].value_counts()

CASH_OUT    2237500
PAYMENT     2151495
CASH_IN     1399284
TRANSFER     532909
DEBIT         41432
Name: type, dtype: int64

## Insights 

- Cash withdrawals are the highest form of transactions made out of all the types of transactions.
- Debit card transactions are the lowest form of transactions.


In [10]:
df[df['step']==2].head()

step     type   amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
2708     2  PAYMENT  1826.81    C70165127        24622.0        22795.19   
2709     2  PAYMENT  7709.20   C520830206        52152.0        44442.80   
2710     2  PAYMENT  3023.16  C1705281026          178.0            0.00   
2711     2  PAYMENT  3708.43   C862062296            0.0            0.00   
2712     2  PAYMENT  8702.16   C521270862        20929.0        12226.84   

         nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
2708  M2026706491             0.0             0.0        0               0  
2709   M351216770             0.0             0.0        0               0  
2710  M1967667267             0.0             0.0        0               0  
2711  M1589968642             0.0             0.0        0               0  
2712   M724525007             0.0             0.0        0               0

In [11]:
df.tail(10)

step      type      amount     nameOrig  oldbalanceOrg  \
6362610   742  TRANSFER    63416.99   C778071008       63416.99   
6362611   742  CASH_OUT    63416.99   C994950684       63416.99   
6362612   743  TRANSFER  1258818.82  C1531301470     1258818.82   
6362613   743  CASH_OUT  1258818.82  C1436118706     1258818.82   
6362614   743  TRANSFER   339682.13  C2013999242      339682.13   
6362615   743  CASH_OUT   339682.13   C786484425      339682.13   
6362616   743  TRANSFER  6311409.28  C1529008245     6311409.28   
6362617   743  CASH_OUT  6311409.28  C1162922333     6311409.28   
6362618   743  TRANSFER   850002.52  C1685995037      850002.52   
6362619   743  CASH_OUT   850002.52  C1280323807      850002.52   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
6362610             0.0  C1812552860            0.00            0.00        1   
6362611             0.0  C1662241365       276433.18       339850.17        1   
6362612             0.0  C1470998563            0.00            0.00        1   
6362613             0.0  C1240760502       503464.50      1762283.33        1   
6362614             0.0  C1850423904            0.00            0.00        1   
6362615             0.0   C776919290            0.00       339682.13        1   
6362616             0.0  C1881841831            0.00            0.00        1   
6362617             0.0  C1365125890        68488.84      6379898.11        1   
6362618             0.0  C2080388513            0.00            0.00        1   
6362619             0.0   C873221189      6510099.11      7360101.63        1   

         isFlaggedFraud  
6362610               0  
6362611               0  
6362612               0  
6362613               0  
6362614               0  
6362615               0  
6362616               0  
6362617               0  
6362618               0  
6362619               0

In [12]:
df['step'].value_counts(ascending = False)

19     51352
18     49579
187    49083
235    47491
307    46968
       ...  
432        4
706        4
693        4
112        2
662        2
Name: step, Length: 743, dtype: int64

## Insights

- There are a total of 743 steps which means that the data is for about 31 days(~ 1 month duration)

In [13]:
transaction_types = df['type'].value_counts().reset_index()

transaction_types.columns = ['Transaction Type', 'Count']

fig = px.bar(
    transaction_types, x = "Transaction Type", y = "Count", text = 'Count',color = "Transaction Type",title = 'Distribution of Transaction type', template = 'plotly_white'
)

fig.update_traces(texttemplate = '%{text:,}',textposition = 'outside')

fig.update_layout(
    xaxis_title = "Transaction Type",yaxis_title = "Count of Transactions",uniformtext_minsize = 8,uniformtext_mode = 'hide',  yaxis = dict(range=[0,transaction_types['Count'].max()* 1.1]))


fig.show()


In [14]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

## Count of Fraud vs non Fraud

In [15]:
fraud_counts = df['isFraud'].value_counts().reset_index()
fraud_counts.columns = ['isFraud','Count']

fraud_counts['isFraud'] = fraud_counts['isFraud'].map({0: 'Non-Fraud', 1 : 'Fraud'})

fig = px.bar(
    fraud_counts, x = 'isFraud', y = 'Count', text = 'Count', color = 'isFraud', title = 'Fraud vs Non-Fraud transactions',template = 'plotly_white')

fig.update_traces(texttemplate = '%{text:,}', textposition = 'outside')

fig.update_layout(xaxis_title = 'Class',yaxis_title = 'Count of Transactions',
                  yaxis = dict(range=[0,fraud_counts['Count'].max()* 1.1]))

fig.show()

In [16]:
total = fraud_counts['Count'].sum()
fraud_counts['Percentage'] = (fraud_counts['Count']/ total * 100).round(2)

fraud_counts

isFraud    Count  Percentage
0  Non-Fraud  6354407       99.87
1      Fraud     8213        0.13

## Insights

- This is an example of an imbalanced dataset since the number of Fraud transactions are quite lower than non fraud & non fraud seems to be extremely high. It is also understandable that the dataset would be set up this way since the number of fraud transactions would be rare. 
- Finding accuracy here would not be a right metric to use. We should explore precision, recall , F1 score etc which i will work on in this project.

In [17]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

## Analyzing fraud distribution by transaction type

In [18]:
fraud_by_type = df.groupby(['type','isFraud']).size().reset_index(name= 'Count')

fraud_by_type['isFraud'] = fraud_by_type['isFraud'].map({0:"Non-Fraud",1: "Fraud"})

fig = px.bar(
    fraud_by_type, x = 'type', y = 'Count',
    color = 'isFraud',barmode = 'group',text = 'Count',
    title = 'Fraud vs non-Fraud Transaction type', template = 'plotly_white')

fig.update_traces(texttemplate = '%{text:,}',textposition = 'outside')

fig.update_layout(xaxis_title = 'Type', yaxis = dict(title = 'Count',
    range=[0,fraud_by_type['Count'].max()* 1.1]))

fig.show()

In [19]:
fraud_by_type

type    isFraud    Count
0   CASH_IN  Non-Fraud  1399284
1  CASH_OUT  Non-Fraud  2233384
2  CASH_OUT      Fraud     4116
3     DEBIT  Non-Fraud    41432
4   PAYMENT  Non-Fraud  2151495
5  TRANSFER  Non-Fraud   528812
6  TRANSFER      Fraud     4097

## Insights

Fraud is observed only in two types:
- CASH_OUT ( ~ 4.1k fraud cases).
- TRANSFER ( ~ 4.1k fraud cases).
- PAYMENT, CASH_IN, DEBIT transactions never appear as fraudulent in this dataset.

In [20]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

In [21]:
fraud_cases = df[df['isFraud'] == 1]

fraud_cases.head(10)

step      type      amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
2        1  TRANSFER      181.00  C1305486145         181.00             0.0   
3        1  CASH_OUT      181.00   C840083671         181.00             0.0   
251      1  TRANSFER     2806.00  C1420196421        2806.00             0.0   
252      1  CASH_OUT     2806.00  C2101527076        2806.00             0.0   
680      1  TRANSFER    20128.00   C137533655       20128.00             0.0   
681      1  CASH_OUT    20128.00  C1118430673       20128.00             0.0   
724      1  CASH_OUT   416001.33   C749981943           0.00             0.0   
969      1  TRANSFER  1277212.77  C1334405552     1277212.77             0.0   
970      1  CASH_OUT  1277212.77   C467632528     1277212.77             0.0   
1115     1  TRANSFER    35063.63  C1364127192       35063.63             0.0   

         nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
2      C553264065             0.0            0.00        1               0  
3       C38997010         21182.0            0.00        1               0  
251    C972765878             0.0            0.00        1               0  
252   C1007251739         26202.0            0.00        1               0  
680   C1848415041             0.0            0.00        1               0  
681    C339924917          6268.0        12145.85        1               0  
724    C667346055           102.0      9291619.62        1               0  
969    C431687661             0.0            0.00        1               0  
970    C716083600             0.0      2444985.19        1               0  
1115  C1136419747             0.0            0.00        1               0

## Feature Engineering

- Adding a few features that can help me understand this dataset better in terms of old balance and new balance.

In [22]:
df['delta_org'] = df['oldbalanceOrg'] - df['newbalanceOrig']
df['delta_dest'] = df['newbalanceDest'] - df['oldbalanceDest']

In [23]:
df['mismatch_org'] = (abs(df['delta_org'] - df['amount']) > 1e-6).astype(int)

df['mismatch_dest'] = (abs(df['delta_dest'] - df['amount']) > 1e-6).astype(int)

In [24]:
df[df['isFraud']== 1].shape

(8213, 15)

In [25]:
df[df['isFraud']== 1].head(10)

step      type      amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
2        1  TRANSFER      181.00  C1305486145         181.00             0.0   
3        1  CASH_OUT      181.00   C840083671         181.00             0.0   
251      1  TRANSFER     2806.00  C1420196421        2806.00             0.0   
252      1  CASH_OUT     2806.00  C2101527076        2806.00             0.0   
680      1  TRANSFER    20128.00   C137533655       20128.00             0.0   
681      1  CASH_OUT    20128.00  C1118430673       20128.00             0.0   
724      1  CASH_OUT   416001.33   C749981943           0.00             0.0   
969      1  TRANSFER  1277212.77  C1334405552     1277212.77             0.0   
970      1  CASH_OUT  1277212.77   C467632528     1277212.77             0.0   
1115     1  TRANSFER    35063.63  C1364127192       35063.63             0.0   

         nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  \
2      C553264065             0.0            0.00        1               0   
3       C38997010         21182.0            0.00        1               0   
251    C972765878             0.0            0.00        1               0   
252   C1007251739         26202.0            0.00        1               0   
680   C1848415041             0.0            0.00        1               0   
681    C339924917          6268.0        12145.85        1               0   
724    C667346055           102.0      9291619.62        1               0   
969    C431687661             0.0            0.00        1               0   
970    C716083600             0.0      2444985.19        1               0   
1115  C1136419747             0.0            0.00        1               0   

       delta_org  delta_dest  mismatch_org  mismatch_dest  
2         181.00        0.00             0              1  
3         181.00   -21182.00             0              1  
251      2806.00        0.00             0              1  
252      2806.00   -26202.00             0              1  
680     20128.00        0.00             0              1  
681     20128.00     5877.85             0              1  
724         0.00  9291517.62             1              1  
969   1277212.77        0.00             0              1  
970   1277212.77  2444985.19             0              1  
1115    35063.63        0.00             0              1

In [26]:
df.groupby('isFraud')[['mismatch_org','mismatch_dest']].mean()*100

mismatch_org  mismatch_dest
isFraud                             
0           80.660666      68.614491
1            0.547912      59.417996

## Insights

In this dataset non-fraud transactions show a higher mismatch rate than fraud transactions. This is because the company - PaySim cancels balances in fraud events, making balance-related features unreliable. For modeling, i would exclude these columns to avoid leakage.

In [27]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  \
0  M1979787155             0.0             0.0        0               0   
1  M2044282225             0.0             0.0        0               0   
2   C553264065             0.0             0.0        1               0   
3    C38997010         21182.0             0.0        1               0   
4  M1230701703             0.0             0.0        0               0   

   delta_org  delta_dest  mismatch_org  mismatch_dest  
0    9839.64         0.0             0              1  
1    1864.28         0.0             0              1  
2     181.00         0.0             0              1  
3     181.00    -21182.0             0              1  
4   11668.14         0.0             0              1

## Adding a new feature called Day of Month

In [28]:
df['day_of_month'] = (df['step'] // 24) + 1

In [29]:
df[['step','day_of_month']].head(30)

step  day_of_month
0      1             1
1      1             1
2      1             1
3      1             1
4      1             1
5      1             1
6      1             1
7      1             1
8      1             1
9      1             1
10     1             1
11     1             1
12     1             1
13     1             1
14     1             1
15     1             1
16     1             1
17     1             1
18     1             1
19     1             1
20     1             1
21     1             1
22     1             1
23     1             1
24     1             1
25     1             1
26     1             1
27     1             1
28     1             1
29     1             1

In [30]:
df['step'].unique()[:50]   # see the first 50 unique step values


array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50])

## Checking if the feature works

In [31]:
df[df['step']== 24].head()

step     type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
571039    24  PAYMENT  27003.49  C1828598998         324.00            0.00   
571040    24  PAYMENT   5655.98  C1782546301       30307.00        24651.02   
571041    24  PAYMENT  10791.26   C672054527      325916.00       315124.74   
571042    24  PAYMENT  24449.53  C1049758403      315124.74       290675.21   
571043    24  PAYMENT  11368.09   C229840744      557128.00       545759.91   

           nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  \
571039   M387058580             0.0             0.0        0               0   
571040   M902526484             0.0             0.0        0               0   
571041   M793205953             0.0             0.0        0               0   
571042  M1342677203             0.0             0.0        0               0   
571043   M403174219             0.0             0.0        0               0   

        delta_org  delta_dest  mismatch_org  mismatch_dest  day_of_month  
571039     324.00         0.0             1              1             2  
571040    5655.98         0.0             0              1             2  
571041   10791.26         0.0             0              1             2  
571042   24449.53         0.0             0              1             2  
571043   11368.09         0.0             0              1             2

In [32]:
df[df['step']== 25].head()

step     type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
574255    25  PAYMENT  19161.18  C1954261867         1589.0            0.00   
574256    25  PAYMENT   7725.76   C912521968       189169.0       181443.24   
574257    25  PAYMENT   7655.50  C1635204165       127302.0       119646.50   
574258    25  PAYMENT  15856.99   C336042058       119646.5       103789.51   
574259    25  PAYMENT   1300.98  C1933422673        20559.0        19258.02   

           nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  \
574255  M1085509720             0.0             0.0        0               0   
574256  M1216280274             0.0             0.0        0               0   
574257  M1025130851             0.0             0.0        0               0   
574258   M737162366             0.0             0.0        0               0   
574259   M474215948             0.0             0.0        0               0   

        delta_org  delta_dest  mismatch_org  mismatch_dest  day_of_month  
574255    1589.00         0.0             1              1             2  
574256    7725.76         0.0             0              1             2  
574257    7655.50         0.0             0              1             2  
574258   15856.99         0.0             0              1             2  
574259    1300.98         0.0             0              1             2

In [33]:
df[df['step']== 743].head()

step      type      amount     nameOrig  oldbalanceOrg  \
6362612   743  TRANSFER  1258818.82  C1531301470     1258818.82   
6362613   743  CASH_OUT  1258818.82  C1436118706     1258818.82   
6362614   743  TRANSFER   339682.13  C2013999242      339682.13   
6362615   743  CASH_OUT   339682.13   C786484425      339682.13   
6362616   743  TRANSFER  6311409.28  C1529008245     6311409.28   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
6362612             0.0  C1470998563             0.0            0.00        1   
6362613             0.0  C1240760502        503464.5      1762283.33        1   
6362614             0.0  C1850423904             0.0            0.00        1   
6362615             0.0   C776919290             0.0       339682.13        1   
6362616             0.0  C1881841831             0.0            0.00        1   

         isFlaggedFraud   delta_org  delta_dest  mismatch_org  mismatch_dest  \
6362612               0  1258818.82        0.00             0              1   
6362613               0  1258818.82  1258818.83             0              1   
6362614               0   339682.13        0.00             0              1   
6362615               0   339682.13   339682.13             0              0   
6362616               0  6311409.28        0.00             0              1   

         day_of_month  
6362612            31  
6362613            31  
6362614            31  
6362615            31  
6362616            31

## Analyzing fraud by day

In [34]:
fraud_by_day = df.groupby('day_of_month')['isFraud'].sum().reset_index()

fig = px.line(
    fraud_by_day,x = 'day_of_month',y = 'isFraud', markers = True, text = 'isFraud' , title = 'Fraud Count per day of month' )

fig.update_traces(textposition = 'top center')

fig.update_layout(xaxis_title = 'Day of the month', yaxis_title = 'Fraud Count', template = 'plotly_white')

fig.show()

## Insights

- Fraud is not uniformly distributed across the month.
- This line chart shows fluctuations across these distributions with alot of spikes and dips in Fraud. 
- The peaks indicate high risk periods particularly on the 3rd to 4th day of the month followed by 7-8th and so on.

In [35]:
fraud_by_day = df.groupby('day_of_month')['isFraud'].sum().reset_index()
fraud_by_day.head(10) 


day_of_month  isFraud
0             1      265
1             2      305
2             3      306
3             4      272
4             5      244
5             6      230
6             7      280
7             8      270
8             9      259
9            10      274

In [36]:
top_spikes = fraud_by_day.sort_values(by = 'isFraud',ascending=False).head(5)

top_spikes.head()

day_of_month  isFraud
16            17      316
2              3      306
1              2      305
11            12      296
30            31      282

In [37]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  \
0  M1979787155             0.0             0.0        0               0   
1  M2044282225             0.0             0.0        0               0   
2   C553264065             0.0             0.0        1               0   
3    C38997010         21182.0             0.0        1               0   
4  M1230701703             0.0             0.0        0               0   

   delta_org  delta_dest  mismatch_org  mismatch_dest  day_of_month  
0    9839.64         0.0             0              1             1  
1    1864.28         0.0             0              1             1  
2     181.00         0.0             0              1             1  
3     181.00    -21182.0             0              1             1  
4   11668.14         0.0             0              1             1

In [47]:
spike_days = [17,3,2,12,31]

df_spike = df[df['day_of_month'].isin(spike_days)]


df_spike[df_spike['day_of_month']==17].head(10)


step     type     amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
5558063   384  PAYMENT     856.26  C1313287211       41625.00        40768.74   
5558064   384  PAYMENT    9751.52  C1208543866       40768.74        31017.22   
5558065   384  PAYMENT   46714.49  C1618473877       31017.22            0.00   
5558066   384  PAYMENT    3455.84  C1997766791       11325.00         7869.16   
5558067   384  PAYMENT    6817.78   C171628663       20409.00        13591.22   
5558068   384  PAYMENT    8156.88   C649123967       12090.00         3933.12   
5558069   384  PAYMENT    7738.69  C1554872709       15627.00         7888.31   
5558070   384  PAYMENT    6688.59  C1391141447         963.00            0.00   
5558071   384  PAYMENT    2334.32   C586575505      169480.00       167145.68   
5558072   384  CASH_IN  100390.24   C410476202       11182.00       111572.24   

            nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  \
5558063    M35864941            0.00             0.0        0               0   
5558064   M396796734            0.00             0.0        0               0   
5558065  M2026368735            0.00             0.0        0               0   
5558066  M1435398664            0.00             0.0        0               0   
5558067  M1305243385            0.00             0.0        0               0   
5558068      M581779            0.00             0.0        0               0   
5558069  M1924689743            0.00             0.0        0               0   
5558070   M362826993            0.00             0.0        0               0   
5558071   M319139735            0.00             0.0        0               0   
5558072   C689331267        14883.45             0.0        0               0   

         delta_org  delta_dest  mismatch_org  mismatch_dest  day_of_month  
5558063     856.26        0.00             0              1            17  
5558064    9751.52        0.00             0              1            17  
5558065   31017.22        0.00             1              1            17  
5558066    3455.84        0.00             0              1            17  
5558067    6817.78        0.00             0              1            17  
5558068    8156.88        0.00             0              1            17  
5558069    7738.69        0.00             0              1            17  
5558070     963.00        0.00             1              1            17  
5558071    2334.32        0.00             0              1            17  
5558072 -100390.24   -14883.45             1              1            17

In [48]:
df_spike[df_spike['day_of_month']==17].shape

(421098, 16)

In [49]:
df_spike[df_spike['day_of_month']==2].shape

(452761, 16)

In [51]:
df_spike['isFraud'].value_counts()

0    1229185
1       1505
Name: isFraud, dtype: int64

## Analyzing fraud spike breakdown for these days

In [55]:
df_spike_filtered = df_spike[df_spike['type'].isin(['CASH_OUT','TRANSFER'])]

In [57]:
fraud_spike = df_spike_filtered.groupby(['day_of_month','type','isFraud']).size().reset_index(name='Count')
fraud_spike

day_of_month      type  isFraud   Count
0              2  CASH_OUT        0  164169
1              2  CASH_OUT        1     154
2              2  TRANSFER        0   37914
3              2  TRANSFER        1     151
4              3  CASH_OUT        0     956
5              3  CASH_OUT        1     153
6              3  TRANSFER        0     542
7              3  TRANSFER        1     153
8             12  CASH_OUT        0  121346
9             12  CASH_OUT        1     148
10            12  TRANSFER        0   27685
11            12  TRANSFER        1     148
12            17  CASH_OUT        0  151281
13            17  CASH_OUT        1     158
14            17  TRANSFER        0   35013
15            17  TRANSFER        1     158
16            31  CASH_OUT        1     141
17            31  TRANSFER        1     141

In [59]:
fraud_rate = (df_spike_filtered.groupby(['day_of_month','type'])['isFraud'].mean()*100).reset_index()

fraud_rate


day_of_month      type     isFraud
0             2  CASH_OUT    0.093718
1             2  TRANSFER    0.396690
2             3  CASH_OUT   13.796213
3             3  TRANSFER   22.014388
4            12  CASH_OUT    0.121817
5            12  TRANSFER    0.531743
6            17  CASH_OUT    0.104332
7            17  TRANSFER    0.449234
8            31  CASH_OUT  100.000000
9            31  TRANSFER  100.000000

## Insights

- Day 2 shows us that Fraud % rate for CASH_OUT & TRANSFER were very less <1%.
- Day 3 shows us that there was a significant increase in frauds occuring in CASH OUT & TRANSFER with the rates being 13.7% & 22% respectively.
- Day 12 shows us that the % rate dropped to a lower level <1%.
- Day 17 shows us that the trend was similar.
- Day 31 shows us that every single transaction for CASH_OUT and TRANSFER were fraudulent as they both have a 100% rate.